### ISSUE TITLE
[data grid] Add "Expand/Collapse All" functionality for tree data

### ISSUE THREAD

**Issue by mahammahmood** [mui/mui-x#14137]
### Summary

I know that you are working on Tree Data Selection. I would like you to consider adding a feature to select or unselect all the checkboxes and to expand and collapse all the nodes as well. Could you please give me the tentative date for the launch of the Tree Data Selection feature?

### Examples

Here is the example 
[https://codesandbox.io/p/sandbox/goofy-beaver-dscr3c](url)

![select and expand](https://github.com/user-attachments/assets/e16b79d1-ffef-4414-89cc-6d27444a77f6)

**Search keywords**: Data ,Grid , Premium, Tree, data, Select All, Expand All, Collapse All
**Order ID**: 91010

**Comment by michelengelen** [comment]
Hey @mahammahmood ... the selection feature is actively worked on atm. #13757 
As soon as the PR gets merged you can expect the feature to land in the next release.

As for the expansion feature we can add this to the board for now to open up the discussion on implementation details. 👍🏼

**Comment by genepaul** [comment]
I am also interested in this functionality.  I tried doing something on my own but was unable to get something that was performant with a lot of rows. I had commented [here](https://github.com/mui/mui-x/issues/8505#issuecomment-2178976301) but never received a response on whether I was doing something wrong. I'd be very interested in how this could work as part of the grid.

**Comment by mahammahmood** [comment]
Is this functionality implemented by MUI?

**Comment by michelengelen** [comment]
yes, it is ... it's apparently a bit harder than originally thought, so the PR got not merged yet ... but we are working on it!

**Comment by mahammahmood** [comment]
Thanks for the update! Do you have an estimated time for when it will be completed?

**Comment by michelengelen** [comment]
no, not that i know of ... @MBilalShafi could give a bit more insight if needed

**Comment by MBilalShafi** [comment]
I added a solution in the userland that could be utilized until a built-in solution is available: https://github.com/mui/mui-x/issues/8505#issuecomment-2384976059

 @mahammahmood @genepaul Could you check if it makes sense for you?

**Comment by michelengelen** [comment]
Good news: everything this issue asked for has shipped.

- **Expand / collapse all:** since [v8.13.0](https://github.com/mui/mui-x/releases/tag/v8.13.0), the grid has `apiRef.current.expandAllRows()` and `apiRef.current.collapseAllRows()` (#19697). They work for tree data and row grouping with client-side data, so a toolbar button like in your example is a one-line handler now. The userland workarounds shared in #8505 are no longer needed.
- **Select / unselect all:** the header checkbox covers it, and with [`rowSelectionPropagation`](https://mui.com/x/react-data-grid/row-selection/#selection-propagation) the selection cascades through the nested children.

The related request to expand groups up to a *specific* depth (instead of all or none) stays tracked in #4930 — please upvote there if you need level-based control.

I will close this issue as completed. Comment here if either API does not cover your case on the latest version.
